File: src/main/java/com/back/domain/member/member/entity/Member.java
package com.back.domain.member.member.entity;

import com.back.domain.member.member.enums.Role;
import com.back.domain.member.member.enums.MemberStatus;
import com.back.domain.member.reputation.entity.Reputation;
import com.back.domain.member.reputation.entity.ReputationEvent;
import com.back.domain.member.review.entity.Review;
import com.back.global.exception.ServiceException;
import com.back.global.jpa.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

@Entity
@Getter
@Setter
@Table(name = "users")
@NoArgsConstructor
public class Member extends BaseEntity {
    @Column(unique = true)
    private String username;

    private String password;

    @Column(unique = true)
    private String nickname;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MemberStatus status;

    @Column(unique = true)
    private String apiKey;

    private String profileImgUrl;

    @OneToOne(mappedBy = "member", fetch = FetchType.LAZY)
    private Reputation reputation;

    @OneToMany(mappedBy = "target", fetch = FetchType.LAZY)
    private List<ReputationEvent> targetEvents;

    @OneToMany(mappedBy = "reporter", fetch = FetchType.LAZY)
    private List<ReputationEvent> reporterEvents;

    @OneToMany(mappedBy = "member", fetch = FetchType.LAZY)
    private List<Review> reviews;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Role role;

    @Column(nullable = true)
    private LocalDateTime suspendAt;  // 정지 시각

    @Column(nullable = true)
    private LocalDateTime deleteAt;  // 영구 정지 시각


    private int loginFailCount;
    @Column(nullable = true)
    private LocalDateTime lastLoginFailAt;
    private boolean locked;
    @Column(nullable = true)
    private LocalDateTime lockedUntil;

    // 계정 잠김 관련
    public boolean isLocked() {
        return locked && lockedUntil != null && lockedUntil.isAfter(LocalDateTime.now());
    }

    public void unlockIfExpired() {
        if (locked && lockedUntil.isBefore(LocalDateTime.now())) {
            this.locked = false;
            this.lockedUntil = null;
            this.loginFailCount = 0;
        }
    }

    public void lock() {
        this.locked = true;
    }

    public void resetFailCount() {
        this.loginFailCount = 0;
    }

    public void increaseFailCount() {
        this.loginFailCount++;
    }

    public void updateLastFailAt(LocalDateTime now) {
        this.lastLoginFailAt = now;
    }

    public void lockUntil(LocalDateTime after) {
        this.lockedUntil = after;
    }


    // 생성자
    public Member(int id, String username, String name, Role role) {
        setId(id);
        this.username = username;
        this.nickname = name;
        this.role = role;
    }

    public Member(String username, String password, String nickname) {
        this(username, password, nickname, null);
    }

    public Member(String username, String password, String nickname, String profileImgUrl) {
        this.username = username;
        this.password = password;
        this.nickname = nickname;
        this.status = MemberStatus.ACTIVE;
        this.apiKey = UUID.randomUUID().toString();
        this.profileImgUrl = profileImgUrl;
        this.loginFailCount = 0;
        this.locked = false;
        setRole(Role.USER);
    }


    public MemberStatus getActive() {
        return this.status;
    }

    public String getNickname() {
        return this.status == MemberStatus.WITHDRAWN ? "탈퇴한 회원" : this.nickname;
    }


    public void checkActorCanModify(Member actor) {
        if (!actor.equals(this))
            throw new ServiceException("403-1", "수정권한이 없습니다.".formatted(getId()));
    }


    // 수정 관련
    public void modifyApiKey(String apiKey) {
        this.apiKey = apiKey;
    }

    public void modifyName(String nickname) {
        this.nickname = nickname;
    }

    public void modifyPassword(String password) {
        this.password = password;
    }

    public void modify(String nickname, String profileImgUrl) {
        this.nickname = nickname;
        this.profileImgUrl = profileImgUrl;
    }


    // 계정 활성화 관련
    // 상태 변경 가능한지
    public void changeStatus(MemberStatus target) {
        if (!this.status.canTransitionTo(target)) {
            throw new ServiceException("400-4", "잘못된 상태 변경입니다.");
        }

        this.status = target;
    }

    // 계정 정지
    public void suspend() {
        changeStatus(MemberStatus.SUSPENDED);
        this.suspendAt = LocalDateTime.now();
    }

    // 계정 재활성화
    public void release() {
        changeStatus(MemberStatus.ACTIVE);
        this.suspendAt = null;
    }

    // 계정 영구 정지
    public void banned() {
        changeStatus(MemberStatus.BANNED);
        this.deleteAt = LocalDateTime.now();
    }

    // 계정 탈퇴
    public void withdraw() {
        changeStatus(MemberStatus.WITHDRAWN);
        this.deleteAt = LocalDateTime.now();
    }


    // 계정 인가 관련
    public boolean isAdmin() {
        if (this.role == Role.ADMIN) return true;

        return false;
    }

    public Collection<? extends GrantedAuthority> getAuthorities() {
        return getAuthoritiesAsStringList()
                .stream()
                .map(SimpleGrantedAuthority::new)
                .toList();
    }


    private List<String> getAuthoritiesAsStringList() {
        List<String> authorities = new ArrayList<>();

        authorities.add("ROLE_USER");

        if (isAdmin()) {
            authorities.add("ROLE_ADMIN");
        }

        return authorities;
    }
}

File: src/main/java/com/back/domain/member/member/entity/Member.kt
package com.back.domain.member.member.entity

import com.back.domain.member.member.enums.MemberStatus
import com.back.domain.member.member.enums.Role
import com.back.domain.member.reputation.entity.Report
import com.back.domain.member.reputation.entity.Reputation
import com.back.domain.member.reputation.entity.ReputationEvent
import com.back.domain.member.review.entity.Review
import com.back.global.exception.ServiceException
import com.back.global.jpa.entity.BaseEntity
import jakarta.persistence.*
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import java.time.LocalDateTime
import java.util.*

@Entity
@Table(name = "users")
class Member(
    @Column(unique = true)
    var username: String,

    var password: String? = null,

    @Column(unique = true)
    var nickname: String,

    @Enumerated(EnumType.STRING)
    var role: Role?,

    var profileImgUrl: String?,

) : BaseEntity() {

    // ========================
    // 필드
    // ========================

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: MemberStatus = MemberStatus.ACTIVE

    @Column(unique = true)
    var apiKey: String? = UUID.randomUUID().toString()

    @OneToOne(mappedBy = "member", fetch = FetchType.LAZY)
    var reputation: Reputation? = null

    @OneToMany(mappedBy = "target", fetch = FetchType.LAZY)
    var targetEvents: MutableList<ReputationEvent?> = mutableListOf()

    @OneToMany(mappedBy = "target", fetch = FetchType.LAZY)
    var targetReports: MutableList<Report?> = mutableListOf()

    @OneToMany(mappedBy = "reporter", fetch = FetchType.LAZY)
    var reporterReports: MutableList<Report?> = mutableListOf()

    @OneToMany(mappedBy = "member", fetch = FetchType.LAZY)
    var reviews: MutableList<Review> = mutableListOf()

    @Column(nullable = true)
    var suspendAt: LocalDateTime? = null

    @Column(nullable = true)
    var deleteAt: LocalDateTime? = null

    @Column(nullable = true)
    var lastLoginFailAt: LocalDateTime? = null

    var locked: Boolean = false

    var loginFailCount: Int = 0

    @Column(nullable = true)
    var lockedUntil: LocalDateTime? = null

    // ========================
    // 보조 생성자
    // ========================

    /** SecurityUser용: id까지 세팅 */
    constructor(id: Int, username: String, nickname: String) : this(
        username,
        null,
        nickname,
        null,
        ""
    ) {
        this.id = id
    }

    constructor(id: Int, username: String, nickname: String, role: Role?) : this(
        username,
        null,
        nickname,
        role,
        ""
    ) {
        this.id = id
    }

    constructor(username: String, password: String?, nickname: String) : this(
        username,
        password,
        nickname,
        null,
        null
    )

    constructor(username: String, password: String?, nickname: String, profileImgUrl: String?) : this(
        username,
        password,
        nickname,
        null,
        profileImgUrl
    )

    /** 회원가입용: apiKey 자동 생성, status ACTIVE */
//    constructor(
//        username: String,
//        password: String?,
//        nickname: String,
//        role: Role,
//        profileImgUrl: String?,
//    ) : this(
//        username,
//        password,
//        nickname,
//        role,
//        profileImgUrl
//    ) {
//        this.status = MemberStatus.ACTIVE
//        this.loginFailCount = 0
//        this.apiKey = UUID.randomUUID().toString()
//        this.locked = false
//    }

    // ========================
    // 프로퍼티
    // ========================

    val isLocked: Boolean
        get() = locked && lockedUntil?.isAfter(LocalDateTime.now()) == true

    val isAdmin: Boolean
        get() = role == Role.ADMIN

    fun getName(): String =
        if (status == MemberStatus.WITHDRAWN) "탈퇴한 회원" else nickname

    val authorities: Collection<GrantedAuthority>
        get() = buildList {
            add(SimpleGrantedAuthority("ROLE_USER"))
            if (isAdmin) add(SimpleGrantedAuthority("ROLE_ADMIN"))
        }

    // ========================
    // 잠금 관련
    // ========================

    fun unlockIfExpired() {
        if (locked && lockedUntil?.isBefore(LocalDateTime.now()) == true) {
            locked = false
            lockedUntil = null
            loginFailCount = 0
        }
    }

    fun lock() { locked = true }
    fun resetFailCount() { loginFailCount = 0 }
    fun increaseFailCount() { loginFailCount++ }
    fun updateLastFailAt(now: LocalDateTime?) { lastLoginFailAt = now }
    fun lockUntil(after: LocalDateTime?) { lockedUntil = after }

    // ========================
    // 수정 관련
    // ========================

    fun modifyApiKey(apiKey: String) { this.apiKey = apiKey }
    fun modifyName(nickname: String) { this.nickname = nickname }
    fun modifyPassword(password: String?) { this.password = password }
    fun modify(nickname: String, profileImgUrl: String?) {
        this.nickname = nickname
        this.profileImgUrl = profileImgUrl
    }

    fun checkActorCanModify(actor: Member) {
        if (actor != this) throw ServiceException("403-1", "수정권한이 없습니다.")
    }

    // ========================
    // 상태 변경 관련
    // ========================

    fun changeStatus(target: MemberStatus) {
        if (!status.canTransitionTo(target)) throw ServiceException("400-4", "잘못된 상태 변경입니다.")
        status = target
    }

    fun suspend() { changeStatus(MemberStatus.SUSPENDED); suspendAt = LocalDateTime.now() }
    fun release() { changeStatus(MemberStatus.ACTIVE); suspendAt = null }
    fun banned() { changeStatus(MemberStatus.BANNED); deleteAt = LocalDateTime.now() }
    fun withdraw() { changeStatus(MemberStatus.WITHDRAWN); deleteAt = LocalDateTime.now() }
}
